autoload/ale/highlight.vim

```diff
@@ -57,6 +78,79 @@ function! ale#highlight#RemoveHighlights() abort
     endfor
 endfunction
 
+endif
+
+if s:nvim_api
+
+function! ale#highlight#UpdateHighlights() abort
+    let l:item_list = get(b:, 'ale_enabled', 1) && g:ale_enabled
+    \   ? get(b:, 'ale_highlight_items', [])
+    \   : []
+
+    call ale#highlight#RemoveHighlights()
+
+    for l:item in l:item_list
+        if get(b:, 'ale_highlight_id', -1) == -1
+            " NOTE: This will highlights nothing but will allocate new id
+            let b:ale_highlight_id = nvim_buf_add_highlight(l:item.bufnr, 0, '', 0, 0, -1)
+        endif
+
+        if l:item.type is# 'W'
+            if get(l:item, 'sub_type', '') is# 'style'
+                let l:group = 'ALEStyleWarning'
+            else
+                let l:group = 'ALEWarning'
+            endif
+        elseif l:item.type is# 'I'
+            let l:group = 'ALEInfo'
+        elseif get(l:item, 'sub_type', '') is# 'style'
+            let l:group = 'ALEStyleError'
+        else
+            let l:group = 'ALEError'
+        endif
```

Comment:
Large chunks of code here are identical. Instead of optionally defining one function or the other, separate the code logically into the common parts between the general Vim API and the NeoVim API, and use a check at the top to call one of the two different implementations. You could, for example, build a `List` containing `Dictionary` objects with all of the data to be set, like the `l:group` value and so on, and then use that data for both implementations.